hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileBlock.java

```diff
@@ -0,0 +1,206 @@
+    this.uncompressedSizeWithoutHeader = readInt(
+        byteBuff, startOffsetInBuff + Header.UNCOMPRESSED_SIZE_WITHOUT_HEADER_INDEX);
+  }
+
+  /**
+   * Parses the HFile block header and returns the {@link HFileBlock} instance based on the input.
+   *
+   * @param context           HFile context.
+   * @param byteBuff          Input data.
+   * @param startOffsetInBuff Offset to start parsing.
+   * @return The {@link HFileBlock} instance based on the input.
+   * @throws IOException if the block cannot be parsed.
+   */
+  public static HFileBlock parse(HFileContext context, byte[] byteBuff, int startOffsetInBuff)
+      throws IOException {
+    HFileBlockType blockType = HFileBlockType.parse(byteBuff, startOffsetInBuff);
+    switch (blockType) {
+      case ROOT_INDEX:
+        return new HFileRootIndexBlock(context, byteBuff, startOffsetInBuff);
+      case FILE_INFO:
+        return new HFileFileInfoBlock(context, byteBuff, startOffsetInBuff);
+      case DATA:
+        return new HFileDataBlock(context, byteBuff, startOffsetInBuff);
+      default:
+        throw new IOException(
+            "Parsing of the HFile block type " + blockType + " is not supported");
+    }
+  }
+
+  /**
+   * Returns the number of bytes needed to store the checksums for
+   * a specified data size.
+   *
+   * @param numBytes number of bytes of data.
+   * @return The number of bytes needed to store the checksum values.
+   */
+  static int numBytes(long numBytes) {
+    return numChunks(numBytes, BYTES_PER_CHECKSUM) * HFileBlock.CHECKSUM_SIZE;
+  }
+
+  /**
+   * Returns the number of checksum chunks needed to store the checksums for
+   * a specified data size.
+   *
+   * @param numBytes         number of bytes of data
+   * @param bytesPerChecksum number of bytes in a checksum chunk
+   * @return The number of checksum chunks
+   */
+  static int numChunks(long numBytes, int bytesPerChecksum) {
+    long numChunks = numBytes / bytesPerChecksum;
+    if (numBytes % bytesPerChecksum != 0) {
+      numChunks++;
+    }
+    if (numChunks > Integer.MAX_VALUE / HFileBlock.CHECKSUM_SIZE) {
+      throw new IllegalArgumentException("The number of chunks is too large: " + numChunks);
+    }
+    return (int) numChunks;
+  }
+
+  /**
+   * Allocates a new byte buffer for uncompressed data and returns a new {@link HFileBlock}
+   * instance for the decompressed content.
+   *
+   * @return
+   */
+  public abstract HFileBlock cloneForUnpack();
+
+  public HFileBlockType getBlockType() {
+    return blockType;
+  }
+
+  public byte[] getByteBuff() {
+    return byteBuff;
+  }
+
+  public int getOnDiskSizeWithHeader() {
+    return onDiskSizeWithoutHeader + HFILEBLOCK_HEADER_SIZE;
+  }
+
+  /**
+   * Decodes and decompresses the block content if the block content is compressed.
+   *
+   * @return {@link HFileBlock} instance
+   * @throws IOException upon decoding and decompression error.
+   */
+  public HFileBlock unpack() throws IOException {
```

Comment:
is it cleaner to just have a `void` return type and for the caller to work with the same block object? A class returning a new instance of that class, feels a lil awkward? do you need a builder here?(or alternatively, should we return a new object for no compression case too?)

Reply:
Yes.  Now the unpacking happens within the same instance, without instantiating a new one.